docs2/site/docs/getting-started/field-middleware.md

```diff
@@ -48,3 +52,63 @@ _.FieldMiddleware.Use(next =>
   };
 });
 ```
+
+Also you can implement `IFieldMiddleware` interface in your Field Middleware:
+
+```csharp
+public interface IFieldMiddleware
+{
+  Task<object> Resolve(IResolveFieldContext context, FieldMiddlewareDelegate next);
+}
+```
+
+It doesn’t have to be implemented on your middleware. Then a search will be made for such a method
+with a suitable signature. Nevertheless, to improve performance, it is recommended to implement
+this interface.
+
+The middleware delegate is defined as:
+
+```csharp
+public delegate Task<object> FieldMiddlewareDelegate(IResolveFieldContext context);
+```
+
+## Field Middleware and Dependency Injection
+
+First, you are advised to read the article about [Dependency Injection](Dependency-Injection).
+
+If you use `IFieldMiddlewareBuilder.Use` overloads which accept type parameter (that is,
+those that do not accept a `IFieldMiddleware` instance or a middleware delegate) then your
+middleware creation will be delegated to DI-container. Thus, you can pass any dependencies to
+the Field Middleware constructor, provided that you have registered them correctly in DI.
+
+## Known issues
+
+Perhaps the most important thing with Field Middlewares that you should be aware of is that the
+default `DocumentExecuter` applies middlewares to the schema only once before the schema is
+initialized. After this, calling any `IFieldMiddlewareBuilder.Use` methods has no effect.
+
+Field Middleware, when applying the schema, **modifies** the resolver of each field. Therefore,
+you should be careful when using different lifetimes (singleton, scoped, transient) for your
+GraphTypes, Schema and Field Middleware. You **can** use any of lifetime, but for example in
+case of using singleton lifetime for some GraphType and scoped lifetime for Field Middleware
+and Schema this will cause the middleware to be applied to the same fields multiple times.
+In the case of ASP.NET Core app the resolvers of these fields will be wrapped again on each
+HTTP request to the server.
+
+General recommendations for lifetimes are:
+
+| Schema    | Graph Type | Middleware | Rating | 
+|-----------|------------|------------|--------|
+| singleton | singleton  | singleton  | the safest and the most performant option recommended by default |
+| scoped    | scoped     | singleton  | much less performant option |
+| scoped    | scoped     | scoped     | the least performant option |
+| scoped    | singleton  | scoped     | DO NOT DO THAT! |
```

Comment:
It would seem that `| singleton | singleton | scoped |` is worth mentioning here.  Based on code review (not testing) it would seem to be that the middleware would pull from the root service provider, in essence operating the same as if the middleware was registered as a singleton.  Note that typically the `InvalidOperationException: Cannot resolve scoped service from root provider` exception is thrown, although setting `options.ValidateScopes = false` ignores the error (but does not change the outcome).

If it were supported -- that is, a singleton schema with a scoped middleware -- that tied the middleware correctly to the current scoped service provider, it would seem to have a significant performance increase over having the schema, graph type, and middleware all scoped.  Perhaps I'm straying off-topic, but this could easily be done if the IResolveFieldContext had a Services property such as in PR #1345 -- the middleware would simply use the `context.Services` property instead of `context.Schema.Services`.  Something to think about anyway.

Reply:
Right, ( singleton | singleton | scoped ) will cause `InvalidOperationException`, therefore, I did not mention this option.

> that tied the middleware correctly to the current scoped service provider, it would seem to have a significant performance increase over having the schema, graph type, and middleware all scoped.

Not really. I do not see the point in such a combination, because the middleware is still applied to the schema once so it is equivalent to ( singleton | singleton | singleton ) . Performance increase is primarily determined by the ability of the schema and its types to be singletons.

If the middleware needs scoped dependencies, they can be obtained indirectly through `IHttpContextAccessor` ctor argument or something like that, maybe `Func<T>`. We discussed this in other issues. But I admit that something else can be improved for middlewares.

Reply:
I added a description of your case in the readme.

Reply:
Looks good to me, given the present options